### ISSUE TITLE
vtile.toJSON viable for equality testing across operating systems

### ISSUE THREAD

**Issue by springmeyer** [mapnik/node-mapnik#442]
Up until v3.3.0 it was possible to do `vtile2.toJSON() == vtile2.toJSON()` to test the equality of vector tiles. As one example [tilelive-bridge](https://github.com/mapbox/tilelive-bridge/blob/master/test/test.js#L127-L141) has been doing this and it has worked well to compare results to fixtures.

However starting at v3.3.0 I'm seeing that fixtures created on OS X no longer work on Linux. While the vector tiles are visually the same but the equality tests fail because the geometries are not the same. My hunch is that the simplification improvements in [mapnik-vt](https://github.com/mapbox/mapnik-vector-tile/pull/94) have brought in minor changes to floating point rounding that manifest with different c++ libraries on OS X vs Linux.

The solution may be as simple as using `static_cast<int>(std::floor(val + .5))` instead of just `static_cast<int>(val)`, or it may be more involved.

**Comment by flippmoke** [comment]
I noticed this with @yhahn earlier in the process, the issue comes from the use of boost numerical_cast more then likely here:
https://github.com/mapnik/mapnik/blob/master/include/mapnik/geometry_strategy.hpp#L157-L178
I am not sure how to resolve it, as we shouldn't always floor. We could add another strategy to add `floor` however.

**Comment by springmeyer** [comment]
Created a branch of tl-bridge that uses 3.3.1-alpha, updated fixtures from OS X, and still failing on linux: https://travis-ci.org/mapbox/tilelive-bridge/builds/63426191.

**Comment by flippmoke** [comment]
OH GOD, so much research here:

So... our little rounding is done currently by the `rounding_cast` that I implemented [here](https://github.com/mapnik/mapnik/blob/master/include/mapnik/util/rounding_cast.hpp). 

The basis of this is that it does basically the same as a `numeric_cast` however, provides the ability to round as well. The type of rounding is done by [Float2IntRounding Policy](http://www.boost.org/doc/libs/1_38_0/libs/numeric/conversion/doc/html/boost_numericconversion/numeric_converter_policy_classes.html#boost_numericconversion.numeric_converter_policy_classes.policy_float2introunder) at the end you might note:

```
The rounder policies supplied by this header use math functions floor() and ceil(). The standard versions of these functions are introduced in context by a using directive, so in normal conditions, the standard functions will be used.
```

Even rounding is a little odd perhaps, and perhaps we should switch to a truncating policy instead. If you look at the code for the Round Policy you see:

`typedef mpl::integral_c<std::float_round_style,std::round_to_nearest> round_style ;`

This comes from the [standard library on how to round](http://en.cppreference.com/w/cpp/types/numeric_limits/round_style) based on different [types](http://en.cppreference.com/w/cpp/types/numeric_limits/float_round_style), but it gets more complex then that... So lets continue down this rabbit hole.

For `double` types compilers are supposed to follow [IEEE 754](http://en.wikipedia.org/wiki/IEEE_floating_point). The wonderful part is that they don't (exactly), and GCC and Clang are allowed to do it differently especially when it comes to arithmetic. Here is a [huge explanation of this on stack overflow about numeric casting and percision arithmetic.](http://stackoverflow.com/a/18993765) but what you want to quickly take note is the following:

* There might have been a loss in precision in the doubles PRIOR to being rounded which might be the issue with arithmetic and compiler options
* There might have been a loss of precision due to casting the value to and from a float at some point such that precision isn't properly carried and different casting rules apply different per compiler. 
* Casting to an int MIGHT have different precision rules between different compilers. 

I think some of this might be a difference in -O2 and -O3 for how math is done, but not certain will have to review those flags.

In short, it might have been just luck that it turned out the same before.

![](http://fc08.deviantart.net/fs70/f/2013/224/5/c/facedesk_gif_by_katetls-d684dcx.gif)

**Comment by springmeyer** [comment]
currently thinking here is that we'll punt on this by checking in platform specific fixtures for tl-bridge. This will dodge the need to fully fix this now, while ensuring that any changes in fixtures are still captured (which is key).

**Comment by springmeyer** [comment]
confirmed this is fixed by:

  - using tl-bridge branch with new fixtures created on OS X
  - built node-mapnik from source on ubuntu precise 64 bit at https://github.com/mapnik/node-mapnik/commit/813bf078b44197a4bedb9d273881ca5428e18d31
  - ensured all tests pass on ubuntu and osx

Fixing commit was by @flippmoke but code now in use moved so the final fixing commit was https://github.com/mapnik/clipper/commit/11859d514ab72270e45937e62f411523273ed958
